Fix many math.round() issues including #1071 "Weird Behavior of round"

LeeTibbert opened the pull request:
* The presenting problem was expressed in issue #1071
    "Weird Behavior of Double's round Method". This issue is now fixed.
    MathSuite.scala was updated to prevent its silent recurrence.

  * To get the CI tests to run, Suite.scala from PR #1305 is included
    in this PR.  I will probably have to rebase after #1305 is merged.

  * This is the first in what is likely to be a long series of fixes
    to Math.Scala. Investigation of #1071 revealed a number of related
    problems in round(). These problems occurred in parallel in the
    handling of Double & Float.

  * Limiting consideration to the two flavors of round() there were
    three classes of problems.

  * The first class of problem is that round() did not properly clamp
    values to the range of a Long or Int, as the argument overload
    demanded.

  * The second class of problem is that round() did not properly
    round NaN arguments to +0.0, as the Java API describes.

  * The third and arguably most important class of problem is that
    round() did not properly round() ties to positive Infinity, as
    the Java API describes.

    The prior code used llvm intrinsics, which are documented to
    use the behavior of libm by default.  Unless its current rounding
    mode is explicitly set differently, libm rounds 0.5 ties away
    from zero.

    Changing llvm/C/libm rounding modes is not entirely thread-friendly.
    Open coding the round method, as this PR does, is easier to get
    correct, is thread-friendly, and almost certainly faster at runtime.

  * Extensive tests have been added to the new round() sections of
    MathSuite.scala.

64/32 bit issues:

      None known

Documentation:

      Deserves a release note.

Testing:

	* Built and tested ("test-all") on X86_64 and X86.

densh commented on unit-tests/src/main/scala/tests/Suite.scala:
```diff
@@ -5,16 +5,23 @@ import sbt.testing.{EventHandler, Logger, Status}
 import scala.reflect.ClassTag
 
+final case class AssertionFailed(msg: String) extends Exception(msg)
+
```
We should keep this PR dedicate to the round only. Can you please rebase it to exclude the test infra changes?

LeeTibbert commented:
#### PR changed as requested.  

Also tiny change made to MathSuite.scala so that it does not depend upon #1305 having
been previously merged.  When/if #1305 gets merged, I will remove those two lines so
that the suite can have the full benefit of asserts with messages. Until then, the verbose
tests document the intent of the test.

LeeTibbert commented:
I believe this is ready for merge when time allows.

Now that PR #1305 has been merged, I removed the unsightly workaround.
I also rebased to current mainline.,  4 green tests in CI.

LeeTibbert commented:
@densh Sorry for the merge difficulty. I believe that I have resolved the conflicts. 4 green checkmarks.

densh commented:
LGTM, thanks @LeeTibbert !
